plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -180,42 +205,85 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(int maxSize)
+    {
+        return CompletableFuture.runAsync(() -> {
+            if (!dynamicFilter.equals(DynamicFilter.EMPTY) || limit.isPresent()) {
+                // With dynamic filter or LIMIT clause, we might reach a point where enough splits have already been generated. In that case, stop enqueuing any more splits.
+                scanTasks.takeWhile(t -> !finished)
+                        .map(queue::offer)
+                        .forEachOrdered(MoreFutures::getFutureValue);
+            }
+            else {
+                scanTasks.map(queue::offer)
+                        .forEachOrdered(MoreFutures::getFutureValue);
             }
-            Optional<FileScanTask> scanTask = getNextFileScanTask();
-            scanTask.ifPresent(fileScanTask -> addScanTaskToSplits(dynamicFilterPredicate, splits, fileScanTask));
+        },
+        executor);
+    }
+
+    private Stream<FileAwareScanTask> tasks()
+    {
+        Optional<CloseableIterator<FileScanTask>> scanTaskIterator = planScanTasks();
+        if (scanTaskIterator.isEmpty()) {
+            return Stream.empty();
         }
-        return completedFuture(new ConnectorSplitBatch(splits, isFinished()));
+
+        long targetSplitSize = getSplitSize(session)
+                .map(DataSize::toBytes)
+                .orElseGet(tableScan::targetSplitSize);
+
+        return Streams.stream(scanTaskIterator.get())
+                .map(wholeFileTask -> {
+                    Iterator<FileScanTask> fileTasksIterator;
+                    if (wholeFileTask.deletes().isEmpty() && noDataColumnsProjected(wholeFileTask)) {
+                        fileTasksIterator = List.of(wholeFileTask).iterator();
+                    }
+                    else {
+                        fileTasksIterator = wholeFileTask.split(targetSplitSize).iterator();
+                    }
+                    List<FileAwareScanTask> subTasks = new ArrayList<>();
+                    while (fileTasksIterator.hasNext()) {
+                        FileScanTask scanTask = fileTasksIterator.next();
+                        filterScanTask(scanTask).ifPresent(task ->
+                                subTasks.add(new FileAwareScanTask(task, !fileTasksIterator.hasNext())));
+                    }
+                    boolean fileHasAnyDeletions = !wholeFileTask.deletes().isEmpty();
+                    subTasks.forEach(subTask -> subTask.setParentFileHasAnyDeletion(fileHasAnyDeletions));
+                    return subTasks;
+                })
+                .flatMap(List::stream);
     }
 
-    private void addScanTaskToSplits(TupleDomain<IcebergColumnHandle> dynamicFilterPredicate, List<ConnectorSplit> splits, FileScanTask scanTask)
+    private void addScanTaskToSplits(TupleDomain<IcebergColumnHandle> dynamicFilterPredicate, List<ConnectorSplit> splits, FileAwareScanTask fileAwareScanTask, AtomicBoolean reachedLimit)
```

Comment:
Why don't we return a boolean here instead of updating an input parameter ?

Reply:
Good call